pydicom/framereader.py

```diff
@@ -0,0 +1,1132 @@
+            f"at location {pixel_data_location} but found tag {tag}"
+        )
+
+    if len(bot) < number_of_frames:
+        logger.debug("build Basic Offset Table item")
+        bot = build_encapsulated_basic_offset_table(
+            fp,
+            first_frame_location=first_frame_location,
+            number_of_frames=number_of_frames,
+        )
+    else:
+        fp.seek(first_frame_location, 0)
+
+    return bot
+
+
+def build_encapsulated_basic_offset_table(
+    fp: DicomFileLike, first_frame_location: int, number_of_frames: int
+) -> List[int]:
+    """Builds a Basic Offset Table (BOT) item of an encapsulated Pixel Data
+    element.
+
+    Parameters
+    ----------
+    fp: pydicom.filebase.DicomFileLike
+        Pointer for DICOM PS3.10 file stream positioned at the first byte of
+        the Pixel Data element following the empty Basic Offset Table (BOT)
+    first_frame_location: location of the first frame item
+    number_of_frames: int
+        Total number of frames in the dataset
+
+    Returns
+    -------
+    List[int]
+        Offset of each Frame item in bytes from the first byte of the Pixel
+        Data element following the BOT item
+
+    Note
+    ----
+    Moves the pointer back to the first byte of the Pixel Data element
+    following the BOT item (the first byte of the first Frame item).
+
+    Raises
+    ------
+    IOError
+        When file pointer is not positioned at first byte of first Frame item
+        after Basic Offset Table item or when parsing of Frame item headers
+        fails
+    ValueError
+        When the number of offsets doesn't match the specified number of frames
+
+    """
+    if fp.tell() != first_frame_location:
+        fp.seek(first_frame_location, 0)
+    offset_values = []
+    fragment_offsets = []
+    i = 0
+    while True:
+        frame_position = fp.tell()
+        tag = TupleTag(fp.read_tag())
+        if int(tag) == SequenceDelimiterTag:
+            break
+        if int(tag) != ItemTag:
+            fp.seek(first_frame_location, 0)
+            raise IOError(
+                "Building Basic Offset Table (BOT) failed. Expected tag of "
+                f"Frame item #{i} at position {frame_position}, but found "
+                f"tag {tag}"
+            )
+        length = fp.read_UL()
+        if length % 2:
+            fp.seek(first_frame_location, 0)
+            raise IOError(
+                "Building Basic Offset Table (BOT) failed. "
+                f"Length of Frame item #{i} is not a multiple of 2."
+            )
+        elif length == 0:
+            fp.seek(first_frame_location, 0)
+            raise IOError(
+                "Building Basic Offset Table (BOT) failed. "
+                f"Length of Frame item #{i} is zero."
+            )
+        fragment_offsets.append(frame_position - first_frame_location)
+        first_two_bytes = fp.read(2, True)
+        if not fp.is_little_endian:
+            first_two_bytes = first_two_bytes[::-1]
```

Comment:
I could remove this since it _appears_ that encapsulated should always be little endian, but leaving in case there's some other reason @hackermd included the if statement.

Reply:
Yes, this is probably not needed, but maybe @hackermd can comment on this. Or generally review this PR, for that...